dali/pipeline/operator/sequence_shape.h

```diff
@@ -255,6 +267,103 @@ UnfoldedSliceRange unfolded_slice_range(const TensorVector<Backend> &data, int s
+ * in the `expand_extents`. Then, the i-th sample of the source batch is repeated
+ * `volume(expand_extents[i])` times in the destination batch. As a result, the destination batch
+ * ends up with `expand_extents.num_elements()` samples.
+ *
+ * The TensorVector specialization is implemented in terms of pointers - the underlying data is not copied.
+ * The TensorList specializations do copy data - due to the contiguity requirements.
+ * @{
+ */
+
+inline TensorListShape<> broadcast_sample_shapes(const TensorListShape<> &shape,
+                                                 int num_expanded_samples,
+                                                 const TensorListShape<> &expand_extents) {
+  assert(num_expanded_samples == expand_extents.num_elements());
+  assert(expand_extents.num_samples() == shape.num_samples());
+  TensorListShape<> broadcast(num_expanded_samples, shape.sample_dim());
+  for (int sample_idx = 0, elem_idx = 0; sample_idx < shape.num_samples(); sample_idx++) {
+    const auto &sample_shape = shape[sample_idx];
+    for (int i = 0; i < volume(expand_extents[sample_idx]); i++) {
+      broadcast.set_tensor_shape(elem_idx++, sample_shape);
+    }
+  }
+  return broadcast;
+}
+
+template <typename Backend>
+void broadcast_samples(TensorVector<Backend> &expanded_batch, const TensorVector<Backend> &batch,
+                       int num_expanded_samples, const TensorListShape<> &expand_extents) {
+  assert(num_expanded_samples == expand_extents.num_elements());
+  assert(expand_extents.num_samples() == batch.num_samples());
+  auto expanded_builder = tv_builder_like(expanded_batch, batch, num_expanded_samples);
+  const auto &shape = batch.shape();
+  const auto &type_info = batch.type_info();
+  for (int sample_idx = 0; sample_idx < batch.num_samples(); sample_idx++) {
+    const auto &slice_shape = shape[sample_idx];
+    int num_elements = volume(expand_extents[sample_idx]);
+    uint8_t *ptr =
+        const_cast<uint8_t *>(static_cast<const uint8_t *>(batch.raw_tensor(sample_idx)));
+    for (int sample_slice_idx = 0; sample_slice_idx < num_elements; sample_slice_idx++) {
+      expanded_builder.SetNext({ptr, slice_shape, type_info.size()});
+    }
+  }
+  assert(expanded_builder.NextSampleIdx() == expanded_batch.num_samples());
+}
+
+inline void broadcast_samples(TensorList<GPUBackend> &expanded_batch,
+                              const TensorList<GPUBackend> &batch, int num_expanded_samples,
+                              const TensorListShape<> &expand_extents,
+                              kernels::ScatterGatherGPU &sg, cudaStream_t stream) {
+  assert(expand_extents.num_samples() == batch.num_samples());
+  const auto &shape = batch.shape();
+  auto broadcast_shape = broadcast_sample_shapes(shape, num_expanded_samples, expand_extents);
+  expanded_batch.SetLayout(batch.GetLayout());
+  expanded_batch.Resize(broadcast_shape, batch.type());
```

Comment:
```suggestion
  expanded_batch.set_order(stream);
  expanded_batch.Resize(broadcast_shape, batch.type());
```
Now that we deal with GPU stuff...

Reply:
done